pulsar-client/src/main/java/org/apache/pulsar/client/impl/AutoClusterFailover.java

```diff
@@ -0,0 +1,234 @@
+/**
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+package org.apache.pulsar.client.impl;
+
+import com.google.common.base.Strings;
+import java.net.InetSocketAddress;
+import java.net.Socket;
+import java.util.Timer;
+import java.util.TimerTask;
+import java.util.concurrent.TimeUnit;
+import lombok.extern.slf4j.Slf4j;
+import org.apache.pulsar.client.api.PulsarClient;
+import org.apache.pulsar.client.api.PulsarClientException;
+import org.apache.pulsar.client.api.ServiceUrlProvider;
+
+@Slf4j
+public class AutoClusterFailover implements ServiceUrlProvider {
+    private PulsarClient pulsarClient;
+    private volatile String currentPulsarServiceUrl;
+    private final String primary;
+    private final String secondary;
+    private final long failoverDelayMs;
+    private final long switchBackDelayMs;
+    private final Timer timer;
+    private volatile long primaryFailedTimestamp;
+    private long primaryRecoverTimestamp;
+    private long secondaryFailedTimestamp;
+    private final int timeout = 30_000;
+
+    private AutoClusterFailover(String primary, String secondary, long failoverDelayMs, long switchBackDelayMs) {
+        this.primary = primary;
+        this.secondary = secondary;
+        this.failoverDelayMs = failoverDelayMs;
+        this.switchBackDelayMs = switchBackDelayMs;
+        this.currentPulsarServiceUrl = primary;
+        this.primaryFailedTimestamp = -1;
+        this.primaryRecoverTimestamp = -1;
+        this.secondaryFailedTimestamp = -1;
+        this.timer = new Timer("pulsar-service-provider");
+    }
+
+    @Override
+    public void initialize(PulsarClient client) {
+        this.pulsarClient = client;
+
+        // start to probe primary cluster active or not
+        this.timer.scheduleAtFixedRate(new TimerTask() {
+            @Override
+            public void run() {
+                // current pulsar serviceUrl is primary
+                if (currentPulsarServiceUrl.equals(primary)) {
+                    if (probeAvailable(primary, timeout)) {
```

Comment:
How about failover after a succession of failures (3 for example, and should be configurable) to avoid unnecessary switches?

Reply:
The current implementation is switch by failure delay time, it can control the client failover time. If we use failure times, the failover time will out of control.